pkg/controller/elasticsearch/pdb/reconcile_with_roles.go

```diff
@@ -0,0 +1,443 @@
+	if slices.Contains(dataRoles, role) {
+		return esv1.DataRole
+	}
+	return role
+}
+
+// reconcileRoleSpecificPDBs creates and reconciles PodDisruptionBudgets per nodeSet roles for enterprise-licensed clusters.
+func reconcileRoleSpecificPDBs(
+	ctx context.Context,
+	k8sClient k8s.Client,
+	es esv1.Elasticsearch,
+	statefulSets sset.StatefulSetList,
+	meta metadata.Metadata,
+) error {
+	// Check if PDB is disabled in the ES spec, and if so delete all existing PDBs (both default and role-specific)
+	// that have a proper owner reference.
+	if es.Spec.PodDisruptionBudget != nil && es.Spec.PodDisruptionBudget.IsDisabled() {
+		if err := deleteDefaultPDB(ctx, k8sClient, es); err != nil {
+			return err
+		}
+		return deleteAllRoleSpecificPDBs(ctx, k8sClient, es)
+	}
+
+	// Retrieve the expected list of PDBs.
+	pdbs, err := expectedRolePDBs(es, statefulSets, meta)
+	if err != nil {
+		return fmt.Errorf("while retrieving expected role-specific PDBs: %w", err)
+	}
+
+	// Reconcile and delete unnecessary role-specific PDBs that could have been created
+	// by a previous reconciliation with a different set of StatefulSets.
+	if err := reconcileAndDeleteUnnecessaryPDBs(ctx, k8sClient, es, pdbs); err != nil {
+		return err
+	}
+
+	// Always ensure any existing default PDB is removed.
+	if err := deleteDefaultPDB(ctx, k8sClient, es); err != nil {
+		return fmt.Errorf("while deleting the default PDB: %w", err)
+	}
+
+	return nil
+}
+
+// expectedRolePDBs returns a slice of PDBs to reconcile based on statefulSet roles.
+func expectedRolePDBs(
+	es esv1.Elasticsearch,
+	statefulSets sset.StatefulSetList,
+	meta metadata.Metadata,
+) ([]*policyv1.PodDisruptionBudget, error) {
+	pdbs := make([]*policyv1.PodDisruptionBudget, 0, len(statefulSets))
+
+	// Group StatefulSets by their connected roles.
+	groups := groupBySharedRoles(statefulSets)
+
+	// Create one PDB per group
+	// Maps order isn't guaranteed so process in order of defined priority.
+	for _, roleName := range priority {
+		group, ok := groups[roleName]
+		if !ok {
+			continue
+		}
+		if len(group) == 0 {
+			continue
+		}
+
+		pdb, err := createPDBForStatefulSets(es, roleName, group, statefulSets, meta)
+		if err != nil {
+			return nil, err
+		}
+		if pdb != nil {
+			pdbs = append(pdbs, pdb)
+		}
+	}
+
+	return pdbs, nil
+}
+
+func groupBySharedRoles(statefulSets sset.StatefulSetList) map[esv1.NodeRole][]appsv1.StatefulSet {
+	n := len(statefulSets)
+	if n == 0 {
+		return map[esv1.NodeRole][]appsv1.StatefulSet{}
+	}
+
+	rolesToIndices := make(map[esv1.NodeRole][]int)
+	indicesToRoles := make(map[int]set.StringSet)
+	for i, sset := range statefulSets {
+		// A statefulSet may not be found within the expected resources,
+		// as it could have been recently deleted/renamed
+		// within the spec so we must retrieve the roles from the
+		// labels in the statefulSet, not the expected configuration.
+		roles := getRolesFromStatefulSet(sset)
+
+		if len(roles) == 0 {
```

Comment:
IIUC `getRolesFromStatefulSet` returns an empty results only if we are in the (unexpected) situation where there is no labels on the `Pod`. In the case of coordinating node I believe we have at least `esv1.CoordinatingRole`.